ground/src/main/java/com/google/android/ground/persistence/sync/MediaUploadWorker.kt

```diff
@@ -0,0 +1,173 @@
+ * marked as such. Otherwise, transient failures are considered eligible for retry.
+ */
+@HiltWorker
+class MediaUploadWorker
+@AssistedInject
+constructor(
+  @Assisted context: Context,
+  @Assisted workerParams: WorkerParameters,
+  private val remoteStorageManager: RemoteStorageManager,
+  private val mutationRepository: MutationRepository,
+  private val userMediaRepository: UserMediaRepository,
+) : CoroutineWorker(context, workerParams) {
+
+  private val loiId: String = workerParams.inputData.getString(LOI_ID) ?: ""
+
+  /**
+   * Left-biased combination of two results. Failures propagate. Success will only be returned in
+   * the case that both results are successful.
+   */
+  private fun <T> join(a: kotlin.Result<T>, b: kotlin.Result<T>): kotlin.Result<T> =
+    if (a.isSuccess) b else a
+
+  override suspend fun doWork(): Result = withContext(Dispatchers.IO) { doWorkInternal() }
+
+  /** Performs media uploads for all eligible submission mutations associated with a given LOI. */
+  private suspend fun doWorkInternal(): Result {
+    check(loiId.isNotEmpty()) { "work was queued for an empty location of interest ID" }
+    Timber.d("Starting media upload for LOI: $loiId")
+
+    // Set of mutations eligible for media upload.
+    val mutations =
+      mutationRepository
+        .getMutations(loiId, MutationEntitySyncStatus.MEDIA_UPLOAD_PENDING)
+        .filterIsInstance<SubmissionMutation>()
+        .filter { it.retryCount < MAX_RETRY_COUNT }
+        .map {
+          it.copy(
+            syncStatus = Mutation.SyncStatus.MEDIA_UPLOAD_IN_PROGRESS,
+            retryCount = it.retryCount + 1
+          )
+        }
+    mutationRepository.updateMutations(mutations) // Mark MEDIA_UPLOAD_IN_PROGRESS
+
+    val updatedMutations = mutations.map { uploadSubmissionMediaAndUpdateSubmission(it) }
+    mutationRepository.updateMutations(updatedMutations)
+
+    return when {
+      // The worker proceeds to a fixed point as follows:
+      //   If there are any FAILED or PENDING mutations after the worker operation, we retry.
+      //   During the retry, FAILED mutations will not be processed, while PENDING ones will only be
+      //   reattempted if they satisfy the retry count bound. Eventually, either all of the
+      //   mutations for this LOI will have succeeded, failed, or exceeded the retry count.
+      updatedMutations.any {
+        it.syncStatus == Mutation.SyncStatus.FAILED ||
+          it.syncStatus == Mutation.SyncStatus.MEDIA_UPLOAD_PENDING
+      } -> Result.retry()
+      else -> Result.success()
+    }
+  }
+
+  /**
+   * Attempts to upload all media associated with a given submission. Updates the submission's sync
+   * status depending on whether or the uploads failed or succeeded.
+   */
+  private suspend fun uploadSubmissionMediaAndUpdateSubmission(
+    mutation: SubmissionMutation
+  ): SubmissionMutation {
+    val photoUploadResult = uploadSubmissionPhotos(mutation)
+    if (photoUploadResult.isSuccess)
```

Comment:
Does this simplify things at all?

`uploadSubmissionPhotos(mutation).fold {
  onSuccess:...
  onFailure: ...
}

Reply:
I think so, thanks!